File: apps/cargo-tracker/internal/handling/domain/handling_type.go
// Package domain は Handling Context のドメインモデル（HandlingActivity 集約・値オブジェクト）を提供する。
package domain

import (
	"errors"

	shared "github.com/k2works/case-study-cargo-tracker/apps/cargo-tracker/internal/shared/domain"
)

// ErrUnknownHandlingType は未知の荷役種別を解釈しようとした場合に返される。
var ErrUnknownHandlingType = errors.New("unknown handling type")

// HandlingType は荷役種別を表す値オブジェクト。
type HandlingType string

const (
	// HandlingTypeReceive は受領。
	HandlingTypeReceive HandlingType = "RECEIVE"
	// HandlingTypeLoad は積込。
	HandlingTypeLoad HandlingType = "LOAD"
	// HandlingTypeUnload は荷降し。
	HandlingTypeUnload HandlingType = "UNLOAD"
	// HandlingTypeCustoms は通関。
	HandlingTypeCustoms HandlingType = "CUSTOMS"
	// HandlingTypeClaim は引取。
	HandlingTypeClaim HandlingType = "CLAIM"
)

// ParseHandlingType は文字列から HandlingType を解釈する。
func ParseHandlingType(v string) (HandlingType, error) {
	switch HandlingType(v) {
	case HandlingTypeReceive, HandlingTypeLoad, HandlingTypeUnload, HandlingTypeCustoms, HandlingTypeClaim:
		return HandlingType(v), nil
	default:
		return "", ErrUnknownHandlingType
	}
}

// String は荷役種別のコード表現を返す。
func (t HandlingType) String() string { return string(t) }

// RequiresVoyageNumber は VoyageNumber が必須かを返す（LOAD / UNLOAD が必須）。
func (t HandlingType) RequiresVoyageNumber() bool {
	return t == HandlingTypeLoad || t == HandlingTypeUnload
}

// IsClaimType は引取作業かを返す。
func (t HandlingType) IsClaimType() bool { return t == HandlingTypeClaim }

// ResultingTransportStatus は荷役完了後に遷移する輸送状態を返す。
// CUSTOMS は輸送フェーズを進めないため UNKNOWN（現状態維持）を返す。
func (t HandlingType) ResultingTransportStatus() shared.TransportStatus {
	switch t {
	case HandlingTypeReceive:
		return shared.TransportStatusReceived
	case HandlingTypeLoad:
		return shared.TransportStatusLoaded
	case HandlingTypeUnload:
		return shared.TransportStatusUnloaded
	case HandlingTypeClaim:
		return shared.TransportStatusClaimed
	default:
		return shared.TransportStatusUnknown
	}
}

// Ja は荷役種別の日本語表示を返す。
func (t HandlingType) Ja() string {
	switch t {
	case HandlingTypeReceive:
		return "受領"
	case HandlingTypeLoad:
		return "積込"
	case HandlingTypeUnload:
		return "荷降し"
	case HandlingTypeCustoms:
		return "通関"
	case HandlingTypeClaim:
		return "引取"
	default:
		return string(t)
	}
}

File: apps/cargo-tracker/src/main/java/com/example/cargotracker/tracking/handling/domain/model/HandlingType.java
package com.example.cargotracker.tracking.handling.domain.model;

/**
 * 荷役種別。
 *
 * <p>航海番号の要否と場所の照合先は<strong>種別自身が知っている</strong>
 * （{@code domain-model.md}「荷役妥当性検証のデシジョンテーブル」）。
 * 登録処理に対応表を書き写すと、種別が増えたときに片方だけが更新される。
 */
public enum HandlingType {

    /** 受領。出発港での貨物受領。場所は予約の出発地と照合する。 */
    RECEIVE("受領", false),

    /** 積込。航海への積み込み。場所は旅程の積込港と照合する。 */
    LOAD("積込", true),

    /** 荷降し。航海からの荷降ろし。場所は旅程の荷降港と照合する。 */
    UNLOAD("荷降し", true),

    /** 通関。手続きであり、場所は照合しない。 */
    CUSTOMS("通関", false),

    /** 引取。目的港での貨物引取。場所は予約の目的地と照合する（US16 / IT7）。 */
    CLAIM("引取", false);

    private final String displayName;
    private final boolean requiresVoyageNumber;

    HandlingType(String displayName, boolean requiresVoyageNumber) {
        this.displayName = displayName;
        this.requiresVoyageNumber = requiresVoyageNumber;
    }

    /** 画面に出す日本語名（正典は {@code ui_design.md}「HandlingType 表示ラベル定義」）。 */
    public String displayName() {
        return displayName;
    }

    /**
     * 航海番号が必須か。
     *
     * <p>積込・荷降しは「どの便に対する作業か」が分からないと、誤配の判定にも
     * 追跡の表示にも使えない。
     */
    public boolean requiresVoyageNumber() {
        return requiresVoyageNumber;
    }

    /**
     * 場所の食い違いを<strong>誤配として確定する</strong>種別か。
     *
     * <p>積込・荷降しは旅程そのものからの逸脱であり、貨物は予定と違う船・違う港へ
     * 向かう（{@code domain-model.md} 荷役ビジネスルール 1）。受領・引取の
     * 食い違いは<strong>警告に留める</strong>。港の中の別のゲートで受け取るなど、
     * 業務上あり得る差であり、輸送そのものは予定どおり進む。
     */
    public boolean misroutesOnLocationMismatch() {
        return this == LOAD || this == UNLOAD;
    }
}
